src/librustc/dep_graph/dep_node.rs

```diff
@@ -127,6 +127,11 @@ pub enum DepNode<D: Clone + Debug> {
     TraitItems(D),
     ReprHints(D),
     TraitSelect(Vec<D>),
+
+    // An optional alternative to `TraitSelect` that avoids a heap allocation
+    // in the case where there is a single D. (Note that `TraitSelect` is still
+    // allowed to contain a Vec with a single D.)
```

Comment:
Sure. I don't think the "type parameter" case is *that* much important here, and eventually we probably want to rethink how we handle this anyway.